Add support for a GUI, virtio-input and virtio-gpu devices

jakecorrenti opened the pull request:
Adds command line support for the user to add virtio-gpu and virtio-input devices which also ties into having a good experience with the newly added gui compatability.

Adds command line support for the user to start a graphical application window when their virtual machine starts through the `--gui` flag (assuming the user added a virtio-gpu device as well).

Adds some mutual exclusion during creation of the virtual machine configuration. Without calling `runtime.LockOSThread()`, starting the virtual machine subsequently starting a graphic application window would result in a SIGILL error. Alongside calling `runtime.LockOSThread()`, parallelized the running of the graphic application window and checking the status of the virtual machine. Otherwise, the two wouldn't be able to run simultaneously.

These new features can be testing using a command similar to this:

```bash
$ ./out/vfkit --cpus 2 --memory 2048 \
--bootloader efi,variable-store=/Users/jakecorrenti/efi-variable-store,create \
--device virtio-serial,stdio \
--device virtio-fs,sharedDir=/Users/jakecorrenti,mountTag=dir0 \
--device virtio-blk,path=/Users/jakecorrenti/Downloads/Fedora-Server-dvd-x86_64-38-1.6.iso \
--device virtio-net,nat,mac=72:20:43:d4:38:62 --device virtio-input,keyboard --device virtio-input,pointing --device virtio-gpu --gui
```

gbraad commented on cmd/vfkit/main.go:
```diff
@@ -168,18 +184,35 @@ func runVirtualMachine(vmConfig *config.VirtualMachine, vm *vz.VirtualMachine) e
+	for _, gpuDev := range vmConfig.VirtioGPUDevices() {
+		if gpuDev.UsesGUI {
+			runtime.LockOSThread()
+			err := vm.StartGraphicApplication(900, 600)
```
are these the dimensions?

jakecorrenti replied:
Fixed!

gbraad commented:
ok-to-test

gbraad commented:
/ok-to-test

baude commented on pkg/config/virtio.go:
```diff
@@ -11,6 +11,27 @@ import (
 // The VirtioDevice interface is an interface which is implemented by all virtio devices.
 type VirtioDevice VMComponent
 
+const (
```
maybe it makes sense to break this group of devices into something like graphics.go or display.go.  Dealer's choice.

baude commented on pkg/config/virtio.go:
```diff
@@ -181,6 +206,109 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+			}
+			dev.Device = option.key
+		} else {
+			return fmt.Errorf("Unknown option for virtio-gpu devices: %s", option.key)
```
consider moving this "up" as an if-condition and inverting the condition.   if option.key != VirtGPUDisplaDevice ...

baude commented on pkg/config/virtio.go:
```diff
@@ -181,6 +206,109 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+
+func (dev *VirtioGPU) FromOptions(options []option) error {
+	for _, option := range options {
+		if option.key == VirtioGPUDisplayDevice {
```
if you move up the error condition below, you can remove this conditional.

baude commented on pkg/vf/virtio.go:
```diff
@@ -53,6 +55,81 @@ func (dev *VirtioBlk) AddToVirtualMachineConfig(vmConfig *vzVirtualMachineConfig
+	gpuDeviceConfig.SetScanouts(
+		graphicsScanoutConfig,
+	)
+	if err != nil {
```
leftover?misplaced?

baude commented on pkg/config/virtio.go:
```diff
@@ -181,6 +206,109 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+func (dev *VirtioGPU) validate() error {
+	if dev.Device != VirtioGPUDisplayDevice {
+		return fmt.Errorf("Unknown option for virtio-gpu devices: %s", dev.Device)
+	} else if dev.Device == "" {
```
optionally dump the extra else

baude commented on pkg/config/virtio.go:
```diff
@@ -181,6 +206,109 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+				return fmt.Errorf(fmt.Sprintf("Unexpected value for virtio-input %s option: %s", option.key, option.value))
+			}
+			dev.InputType = option.key
+		} else {
```
move "up" and invert the logic

baude commented on pkg/config/virtio.go:
```diff
@@ -181,6 +206,109 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+func (dev *VirtioInput) validate() error {
+	if dev.InputType != VirtioInputPointingDevice && dev.InputType != VirtioInputKeyboardDevice {
+		return fmt.Errorf("Unknown option for virtio-input devices: %s", dev.InputType)
+	} else if dev.InputType == "" {
```
optionally dump the else

baude commented on pkg/config/virtio.go:
```diff
@@ -11,6 +11,27 @@ import (
+
+// VirtioGPU configures a GPU device, such as the host computer's display
+type VirtioGPU struct {
+	Device  string // currently supports "display" as the device
```
please add json tags so podman can use them later.

jakecorrenti commented on cmd/vfkit/main.go:
```diff
@@ -168,18 +184,35 @@ func runVirtualMachine(vmConfig *config.VirtualMachine, vm *vz.VirtualMachine) e
+	for _, gpuDev := range vmConfig.VirtioGPUDevices() {
+		if gpuDev.UsesGUI {
+			runtime.LockOSThread()
+			err := vm.StartGraphicApplication(900, 600)
```
I was thinking option 2. Give the user the option to specify custom dimensions, but also default to something like 800x600

baude commented:
LGTM

cfergeau commented on pkg/config/virtio.go:
```diff
@@ -181,6 +220,165 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+
+	return &VirtioInput{
+		InputType: inputType,
+	}, nil
```
This could be written as 
```
    dev := &VirtioInput{
		InputType: inputType,
	}
    if err := dev.validate(); err != nil {
        return nil, err
    }
    return dev, nil
```
to avoid repeating that same validation code multiple times.

cfergeau commented on pkg/config/virtio.go:
```diff
@@ -181,6 +220,165 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+// VirtioGPUNewWithGUIAndDimensions creates a new gpu device for the virtual
+// machine with the graphical user interface enabled and user defined
+// dimensions for its resolution.
+func VirtioGPUNewWithGUIAndDimensions(device string, height, width int) (VirtioDevice, error) {
```
I don't think you need all these constructors as only one or 2 are used ? I'd prefer to have just VirtioGPUNew(...) with a set of args which are mandatory, and a few `VirtioGPUSetXxx` if that's needed.

cfergeau commented on pkg/config/virtio.go:
```diff
@@ -181,6 +220,165 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+
+// VirtioGPUNew creates a new gpu device for the virtual machine.
+// The device parameter is the virtio-gpu device that will be added to the machine
+func VirtioGPUNew(device string) (VirtioDevice, error) {
```
I'm not fully clear why this `device` string is required? This does not seem to be used in `Code-Hex/vz`, and it must be `display` in vfkit code?

cfergeau reviewed (commented):
Thanks a lot for your PR, this is going to be a useful addition to vfkit! 
I'm only half way through the review, but here are my first round of comments.

gbraad commented:
please be aware that the git title is too long and gets broken up. best to just write 'gui' instead of the full 'graphical user interface' as I care more about the addition of the last words: virtio-gpu devices

cfergeau commented:
> please be aware that the git title is too long and gets broken up. best to just write 'gui' instead of the full 'graphical user interface' as I care more about the addition of the last words: virtio-gpu devices

Since you bring this up, I would have preferred (at least) 3 separate commits, each doing one specific thing, "add virtio-input support", "add virtio-gpu support", "add gui support". This also helps with the shortlog length ;)

gbraad commented:
> I would have preferred (at least) 3 separate commits,

Didn't want to be so picky at first, but the description talks about the "I added", "I also added", which refers to these, right. Prefer not to see 'I' in a description either. More like 'This adds'

jakecorrenti commented:
> > I would have preferred (at least) 3 separate commits,
> 
> Didn't want to be so picky at first, but the description talks about the "I added", "I also added", which refers to these, right. Prefer not to see 'I' in a description either. More like 'This adds'
> 
> Let's say, for future commits to separate them?

Fixed. I fixed the PR info but forgot to actually change the commit. Also, feel free to be picky. I want this to be a quality contribution

jakecorrenti commented:
> > please be aware that the git title is too long and gets broken up. best to just write 'gui' instead of the full 'graphical user interface' as I care more about the addition of the last words: virtio-gpu devices
> 
> Since you bring this up, I would have preferred (at least) 3 separate commits, each doing one specific thing, "add virtio-input support", "add virtio-gpu support", "add gui support". This also helps with the shortlog length ;)

I can definitely see this being better, I had the habit of squashing everything. I think it would be beneficial to have some sort of `CONTRIBUTING.md` file to have a location for contributors to see what you want for a PR so you don't have to make the same change requests.

gbraad commented:
we prefer small(er) commits as we have been in situations where bisect was required.

cfergeau commented on pkg/config/virtio.go:
```diff
@@ -181,6 +220,122 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+	virtioGPU.VirtioGPUResolution = VirtioGPUResolution{
+		Height: defaultVirtioGPUResolutionHeight,
+		Width:  defaultVirtioGPUResolutionWidth,
+	}
```
fwiw this can be done in a single statement as well:
```
virtioGPU := &VirtioGPU{
		Device:  VirtioGPUDisplayDevice,
		UsesGUI: false,
		VirtioGPUResolution: VirtioGPUResolution{
				Height: defaultVirtioGPUResolutionHeight,
				Width:  defaultVirtioGPUResolutionWidth,
		},
	}
```

I'm fine with your code, no need to change if you prefer your way.

cfergeau commented on cmd/vfkit/main.go:
```diff
@@ -113,10 +117,22 @@ func waitForVMState(vm *vz.VirtualMachine, state vz.VirtualMachineState) error {
+
+	if opts.UseGUI {
+		for _, gpuDev := range vmConfig.VirtioGPUDevices() {
+			if gpuDev.Device == config.VirtioGPUDisplayDevice {
```
I don't understand why this `Device` field is needed? It seems the code would work equally well if the field was removed?

jakecorrenti replied:
My intent behind it was to leave it open to the possibility that additional graphics devices may be supported. But I can remove it and make the assumption that the only graphics device that will be added to the VM will be the display

cfergeau commented on pkg/config/virtio.go:
```diff
@@ -181,6 +220,122 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+	for _, option := range options {
+		if option.key != VirtioGPUDisplayDevice && option.key != VirtioGPUResolutionWidth && option.key != VirtioGPUResolutionHeight {
+			return fmt.Errorf("Unknown option for virtio-gpu devices: %s", option.key)
+		}
```
This can be the `default:` case in the switch/case below.

jakecorrenti replied:
Fixed.

cfergeau commented on pkg/config/virtio.go:
```diff
@@ -181,6 +220,122 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+		}
+	}
+
+	if dev.Width < 1 && dev.Height < 1 {
```
any reason for checking for `< 1` instead of `== 0`?

jakecorrenti replied:
Fixed.

cfergeau commented on cmd/vfkit/main.go:
```diff
@@ -168,18 +184,35 @@ func runVirtualMachine(vmConfig *config.VirtualMachine, vm *vz.VirtualMachine) e
+
+	for _, gpuDev := range vmConfig.VirtioGPUDevices() {
+		if gpuDev.UsesGUI {
+			runtime.LockOSThread()
```
If I'm not mistaken, `runtime.LockOSThread()` was already called at the beginning of the current function, so this is not really required? I agree it does not hurt to have it here as it's the place that really requires it.

EDIT: without the `runtime.LockOSThread()` call at the beginning of this function, vfkit crashes, so it's not only `StartGraphicApplication` which requires this.

cfergeau commented on cmd/vfkit/main.go:
```diff
@@ -59,6 +60,9 @@ func newBootloaderConfiguration(opts *cmdline.Options) (config.Bootloader, error
 
 func newVMConfiguration(opts *cmdline.Options) (*config.VirtualMachine, error) {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
```
This did not seem required in my testing (latest macOS 13 on a m1 macbook), do you remember why you added this?

jakecorrenti replied:
This isn't required, I must have left it there accidentally

cfergeau reviewed (commented):
A bunch of additional comments now that I could look at the virtio-gpu code and test it, but overall looks good, thanks again!

jakecorrenti commented on cmd/vfkit/main.go:
```diff
@@ -113,10 +117,22 @@ func waitForVMState(vm *vz.VirtualMachine, state vz.VirtualMachineState) error {
+	if opts.UseGUI {
+		for _, gpuDev := range vmConfig.VirtioGPUDevices() {
+			if gpuDev.Device == config.VirtioGPUDisplayDevice {
+				gpuDev.UsesGUI = true
```
I didn't make any changes to the VirtioGPU type. The `UsesGUI` attribute is still there for each Virtio-gpu device, and is automatically set to false. It can only be set to true if the user provides the `--gui` flag which is initially checked here.

cfergeau reviewed (commented):
Apart from this `display` comment, should be good to go in, thanks for the changes you've made so far!
